README.md
# Debian packager (PHP)

 * Develop: [![Build Status](https://travis-ci.org/wdalmut/php-deb-packager.svg?branch=develop)](https://travis-ci.org/wdalmut/php-deb-packager)
 * Master : [![Build Status](https://travis-ci.org/wdalmut/php-deb-packager.svg?branch=master)](https://travis-ci.org/wdalmut/php-deb-packager)

A simple debian packager for PHP applications

Get composer:

```
curl -sS http://getcomposer.org/installer | php 
```

Install dependencies and autoloader

```
php composer.phar install
```

Use it:

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

$control = new \wdm\debian\control\StandardFile();
$control
    ->setPackageName("my-package-name")
    ->setVersion("0.1.1")
    ->setDepends(array("php5", "php5-cli", "php5-xsl"))
    ->setInstalledSize(4096)
    ->setMaintainer("Walter Dal Mut", "walter.dalmut@corley.it")
    ->setProvides("my-package-name")
    ->setDescription("My software description");
;

$packager = new \wdm\debian\Packager();

$packager->setOutputPath("/path/to/out");
$packager->setControl($control);

$packager->mount("/path/to/source-conf", "/etc/my-sw");
$packager->mount("/path/to/exec", "/usr/bin/my-sw");
$packager->mount("/path/to/docs", "/usr/share/docs");

//Creates folders using mount points
$packager->run();

//Get the Debian package command
echo $packager->build();
```

**Create the Package**

```
$(php pack.php)
```

## Pre-Post scripts

Optianally you can add script for different hooks

 * pre-install
   * Run pre install
 * post-install
   * Run post install
 * pre-remove
   * Run pre package remove
 * post-remove
   * Run post package remove

Adding scripts

```php
$packager->setPreInstallScript(__DIR__ . '/my-pre-install-script.sh');
$packager->setPostInstallScript(__DIR__ . '/my-post-install-script.sh');
$packager->setPreRemoveScript(__DIR__ . '/my-pre-remove-script.sh');
$packager->setPreRemoveScript(__DIR__ . '/my-post-remove-script.sh');
```

See a script example

```shell
#!/bin/sh
#postinst script for upcloo

set -e

echo "Goodbye Cruel World"

exit 0
```

## Use Yaml files instead the library directly

Just take a look to [wdalmut/php-deb-describe](https://github.com/wdalmut/php-deb-describe)



CHANGELOG.md
# 0.0.13

## BC Breaks

 * The packager will throws a RuntimeException if the output folder already
     exists and it is not empty.

# 0.0.10

## BC Breaks

## Deprecations

 * Method `mount` will be replaced by method `addMount` in the future

# 0.0.4

## BC Breaks

* Build will returns the dpkg command instead print it to the user
* Removed internal autoloader -> use composer as a default



tests/wdm/debian/PackagerTest.php
<?php
namespace wdm\debian;

use org\bovigo\vfs\vfsStream;

class PackagerTest extends \PHPUnit_Framework_TestCase
{
    public function setUp()
    {
        $this->object = new Packager();
    }

    public function testRetrieveTheBuildCommand()
    {
        $this->object->setOutputPath("/tmp");

        $this->assertEquals("dpkg -b /tmp my.deb", $this->object->build("my.deb"));
    }

    public function testRetriveControlFile()
    {
        $control = new control\StandardFile();
        $this->object->setControl($control);

        $this->assertSame($control, $this->object->getControl());
    }

    /**
     * @expectedException RuntimeException
     * @expectedExceptionMessage OUTPUT DIRECTORY MUST BE EMPTY! Something exists, exit immediately!
     */
    public function testOutputFolderIsNotEmpty()
    {
        $root = vfsStream::setup('root');

        mkdir(vfsStream::url('root/tmp'));
        file_put_contents(vfsStream::url('root/tmp/ciao.txt'), "ciao");

        $this->object->setOutputPath(vfsStream::url('root/tmp'));

        $control = new control\StandardFile();
        $this->object->setControl($control);

        $this->object->run();
    }

    public function testOutputFolderExistsButIsEmpty()
    {
        $root = vfsStream::setup('root');

        mkdir(vfsStream::url('root/tmp'));

        $this->object->setOutputPath(vfsStream::url('root/tmp'));

        $control = new control\StandardFile();
        $this->object->setControl($control);

        $this->object->run();
        $command = $this->object->build();

        $this->assertEquals("dpkg -b vfs://root/tmp tmp.deb", $command);
    }

    public function testCreateDebWhenOutputFolderIsMissing()
    {
        $root = vfsStream::setup('root');

        $this->object->setOutputPath(vfsStream::url('root/tmp'));

        $control = new control\StandardFile();
        $this->object->setControl($control);

        $this->object->run();
        $command = $this->object->build();

        $this->assertEquals("dpkg -b vfs://root/tmp tmp.deb", $command);
    }

    public function testCreateDebPackageWithAnotherName()
    {
        $root = vfsStream::setup('root');

        $this->object->setOutputPath(vfsStream::url('root/tmp'));

        $control = new control\StandardFile();
        $this->object->setControl($control);

        $this->object->run();
        $command = $this->object->build("myname.deb");

        $this->assertEquals("dpkg -b vfs://root/tmp myname.deb", $command);
    }
}


tests/wdm/debian/control/StandardFileTest.php
<?php
namespace wdm\debian\control;

class StandardFileTest extends \PHPUnit_Framework_TestCase
{
    private $object;

    public function setUp()
    {
        $this->object = new StandardFile();
    }

    public function testMinimumFile()
    {
        $conf = (string)$this->object;

        $expected = <<<OEF
Version: 0.1
Section: web
Priority: optional
Architecture: all
Essential: no
Installed-Size: 1024
Maintainer: name [email]
Provides: your-company
Description: Your description

OEF;
        $this->assertEquals($expected, $conf);
    }

    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Invalid property 'MyPersonalSection' for this control file.
     */
    public function testFilterMissingOrInvalidProperties()
    {
        $this->object["MyPersonalSection"] = "Test";
    }

    public function testOverwriteConfiguration()
    {
        $this->object["Version"] = "1.0.1";
        $this->object["Section"] = "Software";
        $this->object["Priority"] = "security";
        $this->object["Architecture"] = "x86";
        $this->object["Essential"] = "yes";
        $this->object["Installed-Size"] = "2048";
        $this->object["Maintainer"] = "Walter Dal Mut [walter.dalmut at gmail dot com]";
        $this->object["Provides"] = "Corley SRL";
        $this->object["Description"] = "My Desc";
        $this->object["Depends"] = "php5-cli";
        $this->object["Recommends"] = "php5-curl";

        $conf = (string)$this->object;

        $expected = <<<OEF
Version: 1.0.1
Section: Software
Priority: security
Architecture: x86
Essential: yes
Depends: php5-cli
Recommends: php5-curl
Installed-Size: 2048
Maintainer: Walter Dal Mut [walter.dalmut at gmail dot com]
Provides: Corley SRL
Description: My Desc

OEF;
        $this->assertEquals($expected, $conf);
    }
}


src/wdm/debian/Packager.php
<?php
namespace wdm\debian;

use RuntimeException;
use wdm\debian\control\StandardFile;

class Packager
{
    private $_control;
    private $_mountPoints = array();
    private $_outputPath;

    private $_preInst;
    private $_postInst;
    private $_preRM;
    private $_postRM;

    /**
     * Set the control file
     *
     * This file contains the base package information
     *
     * @param StandardFile $control
     * @return \wdm\debian\Packager
     */
    public function setControl(StandardFile $control)
    {
        $this->_control = $control;
        return $this;
    }

    /**
     * Return the actual control file
     *
     * @return StandardFile
     */
    public function getControl()
    {
        return $this->_control;
    }

    /**
     * Pre install script
     *
     * @param string $path The absolute path of your pre-install script
     */
    public function setPreInstallScript($path)
    {
    	$this->_preInst = $path;
    }

    /**
     * Post install script
     *
     * @param string $path The absolute path of your post-install script
     */
    public function setPostInstallScript($path)
    {
    	$this->_postInst = $path;
    }

    /**
     * Pre remove script
     *
     * @param string $path The absolute path of your pre-remove script
     */
    public function setPreRemoveScript($path)
    {
    	$this->_preRM = $path;
    }

    /**
     * Post remove script
     *
     * @param string $path The absolute path of your post-remove script
     */
    public function setPostRemoveScript($path)
    {
    	$this->_postRM = $path;
    }

    /**
     * @deprecated See addMount instead
     */
    public function mount($sourcePath, $destinationPath)
    {
        return $this->addMount($sourcePath, $destinationPath);
    }

    public function addMount($sourcePath, $destinationPath)
    {
        $this->_mountPoints[$sourcePath] = $destinationPath;
        return $this;
    }

    public function setOutputPath($path)
    {
        $this->_outputPath = $path;
        return $this;
    }

    public function getOutputPath()
    {
        return $this->_outputPath;
    }

    public function run()
    {
        if (file_exists($this->getOutputPath())) {
            $iterator = new \DirectoryIterator($this->getOutputPath());
            foreach ($iterator as $path) {
                if ($path != '.' && $path != '..') {
                    throw new RuntimeException("OUTPUT DIRECTORY MUST BE EMPTY! Something exists, exit immediately!");
                }
            }
        }

        if (!file_exists($this->getOutputPath())) {
            mkdir($this->getOutputPath(), 0777);
        }

        foreach ($this->_mountPoints as $path => $dest) {
            $this->_pathToPath($path, $this->getOutputPath() . DIRECTORY_SEPARATOR . $dest);
        }

        mkdir($this->getOutputPath() . "/DEBIAN", 0777);

        file_put_contents($this->getOutputPath() . "/DEBIAN/control", (string)$this->_control);

        if ($this->_preInst) {
        	$dest = $this->getOutputPath() . "/DEBIAN/preinst";
        	$this->_copy($this->_preInst, $dest);
        	chmod($dest, 0755);
        }

        if ($this->_postInst) {
        	$dest = $this->getOutputPath() . "/DEBIAN/postinst";
        	$this->_copy($this->_postInst, $dest);
        	chmod($dest, 0755);
        }

        if ($this->_preRM) {
        	$dest = $this->getOutputPath() . "/DEBIAN/prerm";
        	$this->_copy($this->_preRM, $dest);
        	chmod($dest, 0755);
        }

        if ($this->_postRM) {
        	$dest = $this->getOutputPath() . "/DEBIAN/postrm";
        	$this->_copy($this->_postRM, $dest);
        	chmod($dest, 0755);
        }

        return $this;
    }

    private function _pathToPath($path, $dest)
    {
        if (is_dir($path)) {
            $iterator = new \DirectoryIterator($path);
            foreach ($iterator as $element) {
                if ($element != '.' && $element != '..') {
                    $fullPath = $path . DIRECTORY_SEPARATOR . $element;
                    if (is_dir($fullPath)) {
                        $this->_pathToPath($fullPath, $dest . DIRECTORY_SEPARATOR . $element);
                    } else {
                        $this->_copy($fullPath, $dest . DIRECTORY_SEPARATOR . $element);
                    }
                }
            }
        } else if (is_file($path)) {
            $this->_copy($path, $dest);
        }
    }

    private function _copy($source, $dest)
    {
        $destFolder = dirname($dest);
        if (!file_exists($destFolder)) {
            mkdir($destFolder, 0777, true);
        }
        copy($source, $dest);
        if (fileperms($source) != fileperms($dest))
            chmod($dest, fileperms($source));
    }

    public function build($debPackageName = false)
    {
        if (!$debPackageName) {
            $debPackageName = basename($this->getOutputPath() . ".deb");
        }

        $command = "dpkg -b {$this->getOutputPath()} {$debPackageName}";

        return $command;
    }
}


src/wdm/debian/IPackager.php
<?php 
namespace wdm\debian;

use wdm\debian\control\StandardFile;

/**
 * 
 * Packager interface
 *
 * @author Walter Dal Mut
 * @package 
 * @license MIT
 *
 * Copyright (C) 2012 Corley S.R.L.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
interface IPackager
{
    public function setControl(StandardFile $control);
    public function mount($sourcePath, $destinationPath);
    
    /**
     * Output path
     * 
     * @param string $path
     */
    public function setOutputPath($path);
    
    /**
     * Create folder structures
     * 
     * @return IPackager
     */
    public function run();
    
    /**
     * Build the package
     */
    public function build($debPackageName = false);
}

src/wdm/debian/control/StandardFile.php
<?php
namespace wdm\debian\control;

/**
 *
 * Representation of Standard control file for Debian packages
 *
 * @author Walter Dal Mut
 * @package
 * @license MIT
 *
 * Copyright (C) 2012 Corley S.R.L.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
class StandardFile
    implements \ArrayAccess
{
    private $_keys = array(
        'Package' => false,
        'Version' => '0.1',
        "Section" => "web",
        "Priority" => "optional",
        "Architecture" => "all",
        "Essential" => "no",
        "Depends" => false,
        "Pre-Depends" => false,
        "Recommends" => false,
        "Suggests" => false,
        "Installed-Size" => 1024,
        "Maintainer" => "name [email]",
        "Conflicts" => false,
        "Replaces" => false,
        "Provides" => "your-company",
        "Description" => "Your description"
    );

    public function setPackageName($name)
    {
        return $this->_setProperty("Package", $name);
    }

    public function setVersion($version)
    {
        return $this->_setProperty("Version", $version);
    }

    public function setSection($section)
    {
        return $this->_setProperty("Section", $section);
    }

    public function setPriority($priority)
    {
        return $this->_setProperty($this["Priority"], $priority);
    }

    public function setArchitecture($arch)
    {
        return $this->_setProperty("Architecture", $arch);
    }

    public function setEssential($essential)
    {
        return $this->_setProperty("Essential", $essential);
    }

    public function setDepends($depends)
    {
        return $this->_setProperty("Depends", $this->_transformList($depends));
    }

    public function setPreDepends($depends)
    {
        return $this->_setProperty("Pre-Depends", $this->_transformList($depends));
    }

    public function setRecommends($depends)
    {
        return $this->_setProperty("Reccommends", $depends);
    }

    public function setSuggests($depends)
    {
        return $this->_setProperty("Suggests", $this->_transformList($depends));
    }

    public function setInstalledSize($size)
    {
        return $this->_setProperty("Installed-Size", $size);
    }

    public function setMaintainer($maintainer, $email = false)
    {
        $email = ($email) ? $email : "---";
        return $this->_setProperty("Maintainer", $maintainer . "[{$email}]");
    }

    public function setConflicts($conflicts)
    {
        return $this->_setProperty("Conflicts", $this->_transformList($conflicts));
    }

    public function setReplaces($replaces)
    {
        return $this->_setProperty("Conflicts", $this->_transformList($replaces));
    }

    public function setProvides($provides)
    {
        return $this->_setProperty("Provides", $provides);
    }

    public function setDescription($description)
    {
        return $this->_setProperty("Description", $description);
    }

    private function _transformList($depends)
    {
        if (is_array($depends)) {
            $depends = implode(", ", $depends);
        } else {
            $depends = $depends;
        }

        return $depends;
    }

    private function _setProperty($key, $value)
    {
        $this[$key] = $value;
        return $this;
    }

    public function offsetExists ($offset) {
        return array_key_exists($offset, $this->_keys);
    }

    public function offsetGet ($offset) {
        if ($this->offsetExists($offset)) {
            return $this->_keys[$offset];
        } else {
            return null;
        }
    }

    public function offsetSet ($offset, $value) {
        if (!$this->offsetExists($offset)) {
            throw new \InvalidArgumentException("Invalid property '{$offset}' for this control file.");
        }
        $this->_keys[$offset] = $value;
    }

    public function offsetUnset ($offset) {
        if ($this->offsetExists($offset)) {
            unset($this->_keys[$offset]);
        }
    }

    /**
     * Control file string representation.
     *
     * @return string The control file
     */
    public function __toString()
    {
        $control = '';
        foreach ($this->_keys as $key => $value) {
            if($value){
                $control .= "{$key}: {$value}" . PHP_EOL;
            }
        }

        return $control;
    }
}
